File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.java
import java.util.*;

class Solution {
    public int solution(int[] cards) {
        
        Set<Integer> visited = new HashSet<>();
        PriorityQueue<Integer> pq = new PriorityQueue<>(Comparator.reverseOrder());
        
        for(int card : cards){
            
            int now = card;
            int len = 0;
            
            while(!visited.contains(now)){
                visited.add(now);
                len++;
                now = cards[now-1];
            }
            
            pq.add(len);
            
        }
        
        return pq.poll()*pq.peek();
    }
}

File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.js
function solution(cards) {
    
    const set = new Set();
    const que = [];
    
    for(let card of cards){
        
        var now = card;
        var len = 0;
        
        while(!set.has(now)){
            set.add(now);
            len++;
            now = cards[now-1];
        }
        
        que.push(len);
        
    }
    
    que.sort((a,b) => b-a);

    return que[0]*que[1];
}